x/btcstaking/keeper/power_dist_change.go

```diff
@@ -0,0 +1,317 @@
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+/* power distribution update */
+
+// UpdatePowerDist updates the voting power table and distribution cache.
+// This is triggered upon each `BeginBlock`
+func (k Keeper) UpdatePowerDist(ctx context.Context) {
+	height := uint64(sdk.UnwrapSDKContext(ctx).HeaderInfo().Height)
+	btcTipHeight := k.GetCurrentBTCHeight(ctx)
+	maxActiveFps := k.GetParams(ctx).MaxActiveFinalityProviders
+
+	// get the power dist cache in the last height
+	dc := k.getVotingPowerDistCache(ctx, height-1)
+	// get all power distribution update events during the previous tip
+	// and the current tip
+	lastBTCTipHeight := k.GetBTCHeightAtBabylonHeight(ctx, height-1)
+	events := k.GetAllPowerDistUpdateEvents(ctx, lastBTCTipHeight, btcTipHeight)
+
+	// if no event exists, then map previous voting power and
+	// cache to the current height
+	if len(events) == 0 {
+		if dc != nil {
+			// map everything in prev height to this height
+			k.recordVotingPowerAndCache(ctx, dc, maxActiveFps)
+		}
+		return
+	}
+
+	if dc == nil {
+		// no BTC staker at the prior height
+		dc = types.NewVotingPowerDistCache()
+	}
+
+	// clear all events that have been consumed in this function
+	defer func() {
+		for i := lastBTCTipHeight; i <= btcTipHeight; i++ {
+			k.ClearPowerDistUpdateEvents(ctx, i)
+		}
+	}()
+
+	// reconcile old voting power distribution cache and new events
+	// to construct the new distribution
+	newDc := k.processAllPowerDistUpdateEvents(ctx, dc, events, maxActiveFps)
+
+	// record voting power and cache for this height
+	k.recordVotingPowerAndCache(ctx, newDc, maxActiveFps)
+	// record metrics
+	k.recordMetrics(ctx, newDc, maxActiveFps)
+}
+
+func (k Keeper) recordVotingPowerAndCache(ctx context.Context, dc *types.VotingPowerDistCache, maxActiveFps uint32) {
+	babylonTipHeight := uint64(sdk.UnwrapSDKContext(ctx).HeaderInfo().Height)
+
+	// set voting power table for this height
+	for i := uint32(0); i < dc.GetNumActiveFPs(maxActiveFps); i++ {
+		fp := dc.FinalityProviders[i]
+		k.SetVotingPower(ctx, fp.BtcPk.MustMarshal(), babylonTipHeight, fp.TotalVotingPower)
+
+	}
+
+	// set the voting power distribution cache of the current height
+	k.setVotingPowerDistCache(ctx, babylonTipHeight, dc)
+}
+
+func (k Keeper) recordMetrics(ctx context.Context, dc *types.VotingPowerDistCache, maxActiveFps uint32) {
+	// number of active FPs
+	numActiveFPs := int(dc.GetNumActiveFPs(maxActiveFps))
+	types.RecordActiveFinalityProviders(numActiveFPs)
+	// number of inactive FPs
+	numInactiveFPs := len(dc.FinalityProviders) - numActiveFPs
+	types.RecordInactiveFinalityProviders(numInactiveFPs)
+	// staked Satoshi
+	// TODO: some wrapper functions between Satoshis and voting power
+	// to make the 1:1 conversion clear
+	stakedSats := uint64(0)
+	for _, fp := range dc.FinalityProviders {
+		stakedSats += fp.TotalVotingPower
+	}
+	numStakedBTCs := float32(stakedSats / SatoshisPerBTC)
+	types.RecordMetricsKeyStakedBitcoins(numStakedBTCs)
```

Comment:
```suggestion
	stakedSats := btcutil.Amount(0)
	for _, fp := range dc.FinalityProviders {
		// 1 unit of voting power equals to 1 sat
		powerToSats := btcutil.Amount(fp.TotalVotingPower)
		stakedSats += powerToSats
	}
	numStakedBTCs := stakedSats.ToBTC()
	types.RecordMetricsKeyStakedBitcoins(float32(numStakedBTCs))
```
Just wanted to kill this TODO 😄